### PR TITLE
Circuit for opcode `BALANCE`

### PR DESCRIPTION
**Pull request by silathdiir**
Spec: https://github.com/privacy-scaling-explorations/zkevm-specs/pull/248

### Summary

1. Update `BALANCE` bus-mapping and add circuit.
2. Fix to lookup `Balance` if account exists, otherwise lookup `NonExisting`.
3. Add a `HashSet` to save non-existent account addresses in both bus-mapping and circuit `ExeStep`s.

~~Not confirm if necessary to constrain "If the given account doesn't exist, then it will push 0 onto the stack instead." in the circuit.
And is there a method (could be used in circuit) to lookup if account is existing? Thanks.~~

~~Blocked by PR https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/907~~

### CODE REVIEW

**Comment by CPerezz** [comment]
Blocked by https://github.com/privacy-scaling-explorations/zkevm-specs/pull/248 which is at the same time blocked by https://github.com/privacy-scaling-explorations/zkevm-specs/issues/249

**Comment by silathdiir** [comment]
Will try to update this PR since https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/907 has been merged.

**Comment by silathdiir** [comment]
Hi @CPerezz , I updated this `BALANCE` circuit with non-existing proofs. Could you help review again when you have time? Thanks.

File: bus-mapping/src/circuit_input_builder/execution.rs
Comment (CPerezz) [inline bus-mapping/src/circuit_input_builder/execution.rs:49]: For now this looks fine. But IMO we will be able to mix both existant and non-existant accounts within a single HashSet and structure.

File: zkevm-circuits/src/evm_circuit/execution/balance.rs
Comment (CPerezz) [inline zkevm-circuits/src/evm_circuit/execution/balance.rs:129]: I'm unsure why do we need `block.randomness` to set the balance. Is it passed directly as an RLC??

**Review by CPerezz — COMMENTED** [review]
Just wondering the balance passed in the assignation of the Opcode circuit impl.

For the rest this looks good. If you cold provide feedback, I'll approve immediately!

File: zkevm-circuits/src/evm_circuit/execution/balance.rs
Comment (silathdiir) [reply zkevm-circuits/src/evm_circuit/execution/balance.rs]: I checked function [table_assignment_aux](https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/zkevm-circuits/src/witness/rw.rs#L334). But it seems that only `value` is needed here, so I replaced it with `value_assignment(block.randomness)`.

And reference function [value_assignment](https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/zkevm-circuits/src/witness/rw.rs#L518), it seems that the argument `randomness` is used in function `random_linear_combine` as below:
```
AccountFieldTag::CodeHash | AccountFieldTag::Balance => {
    RandomLinearCombination::random_linear_combine(value.to_le_bytes(), randomness)
}
```

Comment (silathdiir) [reply zkevm-circuits/src/evm_circuit/execution/balance.rs]: Hi @CPerezz , could you help review this again when you have time? Thanks.

Comment (CPerezz) [reply zkevm-circuits/src/evm_circuit/execution/balance.rs]: Ye sorry @silathdiir I missed this!!!

**Review by CPerezz — APPROVED** [review]
LGTM! I'd want another review from someone at Scroll before merging this! @icemelon or @silathdiir could you ping someone? :smile:

**Comment by silathdiir** [comment]
> LGTM! I'd want another review from someone at Scroll before merging this! @icemelon or @silathdiir could you ping someone? 😄

Yes. I will ping others to help review. Thanks a lot.

File: zkevm-circuits/src/evm_circuit/execution/balance.rs
Comment (DreamWuGit) [inline zkevm-circuits/src/evm_circuit/execution/balance.rs:73]: ` rw_counter: Delta(cb.rw_counter_offset()),` not good style since we don't know exactly rw counter change, suggest to use normal way  ` rw_counter: 7.expr()`

File: bus-mapping/src/circuit_input_builder/execution.rs
Comment (DreamWuGit) [inline bus-mapping/src/circuit_input_builder/execution.rs:49]: most probably do not need this, for circuit side, existing can be check by reading buss mapping record,  for `AccountFieldTag::NonExisting`, no need additional step info.

File: zkevm-circuits/src/witness/step.rs
Comment (DreamWuGit) [inline zkevm-circuits/src/witness/step.rs:38]: same as above comment

File: zkevm-circuits/src/evm_circuit/execution/balance.rs
Comment (DreamWuGit) [inline zkevm-circuits/src/evm_circuit/execution/balance.rs:108]: want to know what reason for `reverse` it

File: bus-mapping/src/circuit_input_builder/execution.rs
Comment (silathdiir) [reply bus-mapping/src/circuit_input_builder/execution.rs]: Yes exactly 🙏, I fixed to get both `balance` and `exists` in circuit function `assign_exec_step` as below:
```
let (balance, exists) = match block.rws[step.rw_indices[5]] {
    Rw::Account {
        field_tag: AccountFieldTag::Balance,
        value,
        ..
    } => (
        RandomLinearCombination::random_linear_combine(
            value.to_le_bytes(),
            block.randomness,
        ),
        true,
    ),
    Rw::Account {
        field_tag: AccountFieldTag::NonExisting,
        ..
    } => (F::zero(), false),
    _ => unreachable!(),
};
```

File: zkevm-circuits/src/witness/step.rs
Comment (silathdiir) [reply zkevm-circuits/src/witness/step.rs]: Deleted.

File: zkevm-circuits/src/evm_circuit/execution/balance.rs
Comment (silathdiir) [reply zkevm-circuits/src/evm_circuit/execution/balance.rs:108]: Try to investigate `ethers_core::types::Address` and code in `eth-types`, it seems that the `Address` is big endian.

https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/eth-types/src/lib.rs#L167

File: zkevm-circuits/src/evm_circuit/execution/balance.rs
Comment (lispc) [inline zkevm-circuits/src/evm_circuit/execution/balance.rs:51]: can the "query_cell" be moved out of the closure? It is some "structure/allocation" info, better not be guarded with "if" branch.

File: zkevm-circuits/src/evm_circuit/execution/balance.rs
Comment (lispc) [inline zkevm-circuits/src/evm_circuit/execution/balance.rs:70]: use select::expr?

**Comment by lispc** [comment]
need to modify https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/3da1888cfdea1f05062477279d9c5088e01fac19/zkevm-circuits/src/witness/step.rs#L184

File: zkevm-circuits/src/evm_circuit/execution/balance.rs
Comment (silathdiir) [reply zkevm-circuits/src/evm_circuit/execution/balance.rs]: Moved out `query_cell` as:
```
let balance = cb.query_cell();
let exists = cb.query_bool();
cb.condition(exists.expr(), |cb| {
    cb.account_read(
        from_bytes::expr(&address.cells),
        AccountFieldTag::Balance,
        balance.expr(),
    );
});
```

File: zkevm-circuits/src/evm_circuit/execution/balance.rs
Comment (silathdiir) [reply zkevm-circuits/src/evm_circuit/execution/balance.rs]: Fixed to:
```
let gas_cost = select::expr(
    is_warm.expr(),
    GasCost::WARM_ACCESS.expr(),
    GasCost::COLD_ACCOUNT_ACCESS.expr(),
);
```

**Comment by silathdiir** [comment]
> need to modify
> 
> https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/3da1888cfdea1f05062477279d9c5088e01fac19/zkevm-circuits/src/witness/step.rs#L184

Sorry for the wrong revert. Fixed to `OpcodeId::BALANCE => ExecutionState::BALANCE` and moved to next one of `ADDRESS`.

**Review by lispc — APPROVED** [review]

**Pull request merged**
